clang/lib/Sema/SemaDecl.cpp

```diff
@@ -8328,35 +8328,47 @@ void Sema::CheckShadow(NamedDecl *D, NamedDecl *ShadowedDecl,
 
   unsigned WarningDiag = diag::warn_decl_shadow;
   SourceLocation CaptureLoc;
-  if (isa<VarDecl>(D) && isa<VarDecl>(ShadowedDecl) && NewDC &&
-      isa<CXXMethodDecl>(NewDC)) {
+  if (isa<VarDecl>(D) && NewDC && isa<CXXMethodDecl>(NewDC)) {
     if (const auto *RD = dyn_cast<CXXRecordDecl>(NewDC->getParent())) {
       if (RD->isLambda() && OldDC->Encloses(NewDC->getLexicalParent())) {
-        if (RD->getLambdaCaptureDefault() == LCD_None) {
-          // Try to avoid warnings for lambdas with an explicit capture list.
+        if (const auto *VD = dyn_cast<VarDecl>(ShadowedDecl)) {
```

Comment:
Structured bindings aren't always VarDecl (But ValueDecl) so I guess we never diagnosed them !

Reply:
Well, If we want to, I would prefer doing that in a separate PR.

Reply:
We currently warn on that actually 
https://compiler-explorer.com/z/d4W37n9Mn

Can you test that we still do?

The rest of the changes LGTM

Reply:
Ok, added.

Reply:
Thanks!
It surprises me that it works, but great!